fix(5925): suppress ValueError due to invalid datetime header

hi-ogawa opened the pull request:
<!-- Thank you for your contribution! -->

## What do these changes do?

Suppress `ValueError` from `datetime.datetime` function in `BaseRequest._http_date`, which is used for parsing a timestamp in request header (`If-Modified-Since`, `If-Unmodified-Since`, and `If-Range`).
According to the [documentation](https://docs.aiohttp.org/en/stable/web_reference.html?highlight=if_modified#aiohttp.web.BaseRequest.if_modified_since), the corresponding request properties (`if_modified_since`, `if_unmodified_since`, and `if_range`) are supposed to return `None` for invalid date, so this PR will make the behavior aligned to the documentation.

## Are there changes in behavior for the user?

<!-- Outline any notable behaviour for the end users. -->

Well... if there are users who want to handle such error by themselves, then they won't be able to see that error anymore, but that seems unlikely.

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->

Fixes https://github.com/aio-libs/aiohttp/issues/5925

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

webknjaz commented:
I wonder if it's the docs that should be updated. Have you checked if the code changed at some point w/o updating the docs? Merging such a change in a minor release may be dangerous for people relying on the de-facto behavior. Plus raising an exception is more pythonic.

hi-ogawa commented:
@webknjaz Thanks for the feedback!

> Have you checked if the code changed at some point w/o updating the docs?

I found a few relevant PRs, so I will list them here.

- https://github.com/aio-libs/aiohttp/pull/386

  This PR introduced `if_modified_since` with `email.utils.parsedate` and `datetime.datetime` parsing logic. Documentation for `if_modified_since` is also added in this PR and it hasn't changed since then.

- https://github.com/aio-libs/aiohttp/pull/2847

  After the code moving around a few times (e.g. [split web_reqrep.py](https://github.com/hi-ogawa/aiohttp/commit/6baa9e7390099dd2d28d5157a38be4641bb45559)), this is the latest change/refactoring around `if_modified_since`. It also introduced `if_unmodified_since` and `if_range`.

So, in summary, I don't think there was any change in behavior w/o updating the docs.

---

> Plus raising an exception is more pythonic.

For this one, I cannot weight in. So I'll leave it to core maintainers for the final decision.
However, I'll just make a pointer to what other libraries do.

- For Flask's werkzeug library, they have similar properties on `Request` class (e.g. [`if_modified_since`](https://github.com/pallets/werkzeug/blob/f7b53ee8a81f015c6c85659959004328237a3c8c/src/werkzeug/sansio/request.py#L442)) and they use a [`parse_date`](https://github.com/pallets/werkzeug/blob/f7b53ee8a8/src/werkzeug/http.py#L921-L947) utility function, which is similar to what aiohttp does, but it suppresses exceptions.


- For Django (personally I'm not familiar with this though), it doesn't seem there are properties like `if_modified_since`, but there is a utility function [`parse_http_date_safe`](https://github.com/django/django/blob/ca9872905559026af82000e46cde6f7dedc897b6/django/utils/http.py#L136) which can parse date header without exception.

---

While grepping around the code base, I actually found `StreamResponse.last_modified`, where same date parsing logic is copy/pasted. So, I will update this as well, possibly with small refactoring.

hi-ogawa commented:
Another thing I noticed is that current aiohttp's `if_modified_since` ignores timezone in the original header (e.g. "GMT", "+0900" etc..) and overwrites with `datetime(... tzinfo=datetime.timezone.utc)`, which might not be the behavior users would expect.
I would suggest to use `email.utils.parsedate_to_datetime` to parse directly from `str` to `datetime` including timezone as werkzeug's [`parse_date`](https://github.com/pallets/werkzeug/blob/f7b53ee8a8/src/werkzeug/http.py#L921-L947) does (this function is available since Python 3.3 so it meets aiohttp's requirement Python >= 3.7).
However, note that currently `email.utils.parsedate_to_datetime` has a bit tricky behavior regarding parse errors (see https://bugs.python.org/issue30681), but we can just use `suppress(TypeError, ValueError)` and it should be fine.

hi-ogawa commented:
With this commit https://github.com/aio-libs/aiohttp/pull/6012/commits/42f1e975057e92bf06aa39cdbcdb9f896ee4a76b, I addresed my own comment https://github.com/aio-libs/aiohttp/pull/6012#issuecomment-922391053.

> While grepping around the code base, I actually found `StreamResponse.last_modified`, where same date parsing logic is copy/pasted. So, I will update this as well, possibly with small refactoring.

Dreamsorcerer commented:
>     * [Make StaticRoute support Last-Modified and If-Modified-Since headers #386](https://github.com/aio-libs/aiohttp/pull/386)
>       This PR introduced `if_modified_since` with `email.utils.parsedate` and `datetime.datetime` parsing logic. Documentation for `if_modified_since` is also added in this PR and it hasn't changed since then.

Looks to me like the original code expected the `parsedate()` function to only return results if they would also be valid for `datetime()`, not realising that it could parse, for example, a year number which is larger than what `datetime()` will accept. So, this looks like a subtle bug in the original implementation.

> > Plus raising an exception is more pythonic.

In general, but these little helper attributes seem like they should not raise exceptions. I don't expect to get a `ValueError` when accessing an attribute. If I did want a more precise control over the parsing with exceptions and such, I'd probably look at the raw header and parse it manually, rather than using this attribute.

Dreamsorcerer commented:
@webknjaz Any other concerns or can we merge this?
